### ISSUE TITLE
Cannot add feed - Cloudflare protection

### ISSUE THREAD

**Issue by ryancom16** [FreshRSS/FreshRSS#1646]
Adding a Feed protected by cloudflare results in a 503 error

 curl -v -A "SimplePie/1.7" https://daz3dfree.com/feed

```

* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
> GET /feed HTTP/1.1
> Host: daz3dfree.com
> User-Agent: SimplePie/1.7
> Accept: */*
>
< HTTP/1.1 503 Service Temporarily Unavailable
< Date: Thu, 28 Sep 2017 01:40:46 GMT
< Content-Type: text/html; charset=UTF-8
< Transfer-Encoding: chunked
< Connection: close
< Set-Cookie: __cfduid=d3a9c091c5a7af5c2a05f2eca56b1c85e1506562846; expires=Fri, 28-Sep-18 01:40:46 GMT; path=/; domain=.daz3dfree.com; HttpOnly
< X-Frame-Options: SAMEORIGIN
< Refresh: 8;URL=/cdn-cgi/l/chk_jschl?pass=1506562850.398-PEHY94aqMK
< Cache-Control: no-cache
< Server: cloudflare-nginx
< CF-RAY: 3a52ff1dfe2b828b-ATL
<
<!DOCTYPE HTML>
<html lang="en-US">
<head>
  <meta charset="UTF-8" />
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1" />
  <meta name="robots" content="noindex, nofollow" />
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" />
  <title>Just a moment...</title>
  <style type="text/css">
    html, body {width: 100%; height: 100%; margin: 0; padding: 0;}
    body {background-color: #ffffff; font-family: Helvetica, Arial, sans-serif; font-size: 100%;}
    h1 {font-size: 1.5em; color: #404040; text-align: center;}
    p {font-size: 1em; color: #404040; text-align: center; margin: 10px 0 0 0;}
    #spinner {margin: 0 auto 30px auto; display: block;}
    .attribution {margin-top: 20px;}
    @-webkit-keyframes bubbles { 33%: { -webkit-transform: translateY(10px); transform: translateY(10px); } 66% { -webkit-transform: translateY(-10px); transform: translateY(-10px); } 100% { -webkit-transform: translateY(0); transform: translateY(0); } }
    @keyframes bubbles { 33%: { -webkit-transform: translateY(10px); transform: translateY(10px); } 66% { -webkit-transform: translateY(-10px); transform: translateY(-10px); } 100% { -webkit-transform: translateY(0); transform: translateY(0); } }
    .bubbles { background-color: #404040; width:15px; height: 15px; margin:2px; border-radius:100%; -webkit-animation:bubbles 0.6s 0.07s infinite ease-in-out; animation:bubbles 0.6s 0.07s infinite ease-in-out; -webkit-animation-fill-mode:both; animation-fill-mode:both; display:inline-block; }
  </style>

    <script type="text/javascript">
  //<![CDATA[
  (function(){
    var a = function() {try{return !!window.addEventListener} catch(e) {return !1} },
    b = function(b, c) {a() ? document.addEventListener("DOMContentLoaded", b, c) : document.attachEvent("onreadystatechange", b)};
    b(function(){
      var a = document.getElementById('cf-content');a.style.display = 'block';
      setTimeout(function(){
        var s,t,o,p,b,r,e,a,k,i,n,g,f, lxVTpdd={"btCDcyKZSc":+((!+[]+!![]+[])+(!+[]+!![]+!![]+!![]+!![]+!![]+!![]))};
        t = document.createElement('div');
        t.innerHTML="<a href='/'>x</a>";
        t = t.firstChild.href;r = t.match(/https?:\/\//)[0];
        t = t.substr(r.length); t = t.substr(0,t.length-1);
        a = document.getElementById('jschl-answer');
        f = document.getElementById('challenge-form');
        ;lxVTpdd.btCDcyKZSc*=!+[]+!![]+!![];lxVTpdd.btCDcyKZSc-=+((!+[]+!![]+[])+(!+[]+!![]+!![]+!![]+!![]+!![]+!![]+!![]));lxVTpdd.btCDcyKZSc-=+((!+[]+!![]+!![]+!![]+[])+(!+[]+!![]+!![]+!![]));a.value = parseInt(lxVTpdd.btCDcyKZSc, 10) + t.length; '; 121'
        f.submit();
      }, 4000);
    }, false);
  })();
  //]]>
</script>


</head>
<body>
  <table width="100%" height="100%" cellpadding="20">
    <tr>
      <td align="center" valign="middle">
          <div class="cf-browser-verification cf-im-under-attack">
  <noscript><h1 data-translate="turn_on_js" style="color:#bd2426;">Please turn JavaScript on and reload the page.</h1></noscript>
  <div id="cf-content" style="display:none">

    <div>
      <div class="bubbles"></div>
      <div class="bubbles"></div>
      <div class="bubbles"></div>
    </div>
    <h1><span data-translate="checking_browser">Checking your browser before accessing</span> daz3dfree.com.</h1>

    <p data-translate="process_is_automatic">This process is automatic. Your browser will redirect to your requested content shortly.</p>
    <p data-translate="allow_5_secs">Please allow up to 5 seconds&hellip;</p>
  </div>

  <form id="challenge-form" action="/cdn-cgi/l/chk_jschl" method="get">
    <input type="hidden" name="jschl_vc" value="9a3614bc707637fa448c62545811e7e2"/>
    <input type="hidden" name="pass" value="1506562850.398-PEHY94aqMK"/>
    <input type="hidden" id="jschl-answer" name="jschl_answer"/>
  </form>
</div>


          <div class="attribution">
            <a href="https://www.cloudflare.com/5xx-error-landing?utm_source=iuam" target="_blank" style="font-size: 12px;">DDoS protection by Cloudflare</a>
            <br>
            Ray ID: 3a52ff1dfe2b828b
          </div>
      </td>

    </tr>
  </table>
</body>
</html>
* TLSv1.2 (OUT), TLS alert, Client hello (1):
```

**Comment by Alkarex** [comment]
Indeed. If someone finds a cURL command that works for this feed, let me know.

**Comment by Alkarex** [comment]
If it can work with special HTTP headers, then it might be fixed with https://github.com/FreshRSS/FreshRSS/issues/1627

**Comment by Alkarex** [comment]
It looks like Cloudflare requires both cookies and JavaScript to work

**Comment by Frenzie** [comment]
I saw https://github.com/KyranRana/cloudflare-bypass which seems a bit complicated. But someone could investigate it in an add-on or something.

**Comment by KyranRana** [comment]
I am rewriting the PHP library so any suggestions are welcome :)

**Comment by Alkarex** [comment]
I am a bit afraid this will become a race towards higher complexity, with Cloudflare making it more and more difficult when attacks start using libraries supporting Cloudflare.
I am not sure to be willing to go down this route: I have a bit the feeling that Cloudflare is "breaking the Web" with their approaches, so it should not be without consequences for Web site owners to activate those barriers.
P.S.: But I would for sure welcome a Cloudflare support as an extension in FreshRSS

**Comment by Frenzie** [comment]
> I have a bit the feeling that Cloudflare is "breaking the Web" with their approaches

That too, though I'm more worried about the fact that a third or so of the web is effectively "hosted" on it. Decentralized, hah! :-P

**Comment by ryancom16** [comment]
Successfully tested [Cloudflare-Bypass](https://github.com/KyranRana/cloudflare-bypass) on cloudflare protected feeds. Hoping to create an extension soon. What are my next steps? In other words, how best would it be implemented?

**Comment by Alkarex** [comment]
Sorry for the delay @ryancom16 
I first need to add an additional deeper API hook to allow overriding the cURL exec method.
The general approach will then be similar than https://github.com/FreshRSS/FreshRSS/blob/master/extensions/Tumblr-GDPR/
We then need an approach to find out when to use this hook or not. Probably an option at feed level. Maybe there are e.g. some HTTP headers in the responses which could give a hint to decide whether the hook should be used or not.

**Comment by Alkarex** [comment]
Example of cURL response for https://coinkite.com , which uses the CloudFlare UAM:

```
< HTTP/1.1 503 Service Temporarily Unavailable
< Date: Mon, 23 Jul 2018 21:45:11 GMT
< Content-Type: text/html; charset=UTF-8
< Transfer-Encoding: chunked
< Connection: close
< X-Frame-Options: SAMEORIGIN
< Set-Cookie: __cfduid=d66e985fbc48942b64e56c6aadd92cd261532382311; expires=Tue, 23-Jul-19 21:45:11 GMT; path=/; domain=.coinkite.com; HttpOnly
< Cache-Control: no-cache
< Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
< Server: cloudflare
< CF-RAY: 43f1552869a23d31-CPH
```

Probably the Status 503 together with `Server: cloudflare` should be enough for detecting the need.

**Comment by julienM77** [comment]
Hey, any update for bypass cloundfare protection ?

**Comment by squromiv** [comment]
I have the same problem with the latest version of FreshRSS. Feeds are available in browser, but can not be added to FreshRSS. Is there any news, resolving this issue?

**Comment by Alkarex** [comment]
> I have the same problem with the latest version of FreshRSS. Feeds are available in browser, but can not be added to FreshRSS. Is there any news, resolving this issue?

Not at the moment, and I am not working on it personally. If you can make it to work from cURL and if it is valid according to https://validator.w3.org/feed/ , we should be able to support it in FreshRSS. If not, I think the most reasonable is to contact the Web site owners to ask them to unbreak their feed. As last option, such workarounds should be in an extension.

**Comment by ryancom16** [comment]
@squromiv Lately I've been using [FlareSolverr](https://github.com/FlareSolverr/FlareSolverr). I spun up a Flaresolverr docker container and then use a simple php file I created to take in the feed address as a parameter, create the request to Flaresolverr, and return the response.

**Comment by squromiv** [comment]
@Alkarex 
Please, take a look at [this](https://www.abw.by/rss/all.rss) feed.

**Comment by Frenzie** [comment]
@squromiv 
The feed doesn't seem to do anything weird for me. If it's behind CloudFlare maybe it misdetected your IP as a spammer or something.

@Alkarex 
> I think the most reasonable is to contact the Web site owners to ask them to unbreak their feed.

+1

Feeds aren't made to be fetched by full-featured web browsers, so basically you've just got a broken feed there and nothing more.

@ryancom16 
> Lately I've been using FlareSolverr.

Heh, I was going to jokingly suggest you could always use Selenium to browse to the website and use its cookies. ;-)

**Comment by squromiv** [comment]
> maybe it misdetected your IP as a spammer or something

The feed is opened in browser, but can not be added in Freshrss.
![screen_2021 09 12](https://user-images.githubusercontent.com/15985877/132997662-0aa8bec7-a554-47c5-898d-ef1fb0a4f281.png)

**Comment by Frenzie** [comment]
Your FreshRSS instance is running on the same IP? But yes, I'm not talking about my browser.

![Screenshot 2021-09-12 at 19-55-38 (67) Автоновости от ABW BY · FreshRSS](https://user-images.githubusercontent.com/202757/132997797-d99d4220-c622-403a-b436-69606253d3c1.png)

**Comment by squromiv** [comment]
> Your FreshRSS instance is running on the same IP?

Yes.
What version of Freshrss are you using?

**Comment by Frenzie** [comment]
1.18.1 and latest edge (potentially give or take a few commits)

**Comment by Alkarex** [comment]
It seems to be based on IP plus other HTTP headers.
This returns some HTML+JavaScript from my home address and OVH server:

```
curl -v -H 'Accept: application/atom+xml, application/rss+xml, application/rdf+xml;q=0.9, application/xml;q=0.8, text/xml;q=0.7' 'https://www.abw.by/rss/all.rss'
```

**Comment by squromiv** [comment]
> It seems to be based on IP plus

Server and browser are placed on the same machine. The feed is available in browser, but can not be added to a reader. 
![screen_2021 09 12](https://user-images.githubusercontent.com/15985877/132998824-790d14b4-8db0-4f3f-850b-a3f6d2f09c67.png)

**Comment by Alkarex** [comment]
> > It seems to be based on IP plus
> 
> Server and browser are placed on the same machine. The feed is available in browser, but can not be added to a reader.
> ![screen_2021 09 12](https://user-images.githubusercontent.com/15985877/132998824-790d14b4-8db0-4f3f-850b-a3f6d2f09c67.png)

Yes, because it requires JavaScript... except for some IPs / HTTP headers

**Comment by squromiv** [comment]
So how to resolve the issue?

**Comment by Alkarex** [comment]
@squromiv See https://github.com/FreshRSS/FreshRSS/issues/1646#issuecomment-917671995
It is not a problem on FreshRSS side

**Comment by squromiv** [comment]
> Lately I've been using [FlareSolverr](https://github.com/FlareSolverr/FlareSolverr). I spun up a Flaresolverr docker container and then use a simple php file I created to take in the feed address as a parameter, create the request to Flaresolverr, and return the response.

Can you share the file? And how to use it with Freshrss?

**Comment by keresztg** [comment]
I'm having a similar problem with https://media1.hu/feed/: `curl` and `wget` fails miserably with 503 and some weird content.
But from the very same machine `w3m` and `links` can load the feed properly.
So I've created a cron job which dumps the feed with `w3m` into a folder of my HTTP server, something like:
```
*/5 * * * * cd /var/www/html && w3m -dump https://media1.hu/feed/ > media1-rss.xml
```
And I use this own URL in FreshRSS.
YMMV.

**Comment by Alkarex** [comment]
@keresztg Thanks for this good idea of workaround. If this proves to be a more general solution, we could add a function in FreshRSS to call an external program for downloading a feed.

**Comment by Frenzie** [comment]
I'm surprised w3m manages to get around the script. Perhaps running curl/wget with the w3m user agent would yield the same result?

**Comment by keresztg** [comment]
@Frenzie Altering the user agent on the curl command did not help me out. Tried and failed with the following commands:
```
curl -A 'w3m/0.5.3+git20190105' https://media1.hu/feed/
curl -A 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:92.0) Gecko/20100101 Firefox/92.0' https://media1.hu/feed/
```
So something else also happens inside w3m and links... Maybe some cookie and redirect handling?

**Comment by keresztg** [comment]
@Alkarex I would be a little bit hesitant about calling arbitrary executables from a PHP... From security perspective that would open a new can of worms.
Maybe a new FAQ entry would be enough for the average technologically advanced users.

**Comment by Frenzie** [comment]
> Maybe some cookie and redirect handling?

It's possible that depending on the specific tool in question you might need to explicitly enable support for session cookies and/or redirects, but I'm reasonably sure they both support it. Curl might not support something like meta-tag based redirects, potentially?

**Comment by squromiv** [comment]
> So I've created a cron job which dumps the feed with `w3m`

Really? I have the same bad result with w3m, as well as with curl.

**Comment by keresztg** [comment]
@squromiv Well, I can't rule out that CloudFlare has multiple levels of ~~annoyance~~ protection...
One of my IPs produce the same result: also w3m and links fails with https://www.abw.by/rss/all.rss
Only shows the page which tell that "Please turn JavaScript on and reload the page. Checking your browser before accessing abw.by. Please enable Cookies and reload the page."
Really seems to me that it won't work without JavaScript.

**Comment by squromiv** [comment]
> the same result: also w3m and links fails

Exactly.
![screen_2021 09 28](https://user-images.githubusercontent.com/15985877/135154395-55b4d471-9172-4e46-9be6-ee3af02ceb32.png)

**Comment by lukerg** [comment]
I think I am able to offer some serious relief to many here and for others. It would appear to be the case that cloudflare is profiling HTTPS clients by the TLS/SSL options negotiated. I have found through repeated experiementation that instructing curl et al to NOT send ALPN, along with a spoofed user-agent header results in the request going through. It also explains the difference between elinks/w3m/links working yet curl & wget failing despite spoofing the user-agent header.
I was initially worried the profiling was more arcane, such as the order and quantity of SSL ciphers offered; cloudflare (or someone else) published a research paper about that very method, which is what started me down this path.

To the point, modifying line 371of Entry.php to include 
` CURLOPT_SSL_ENABLE_ALPN => false,`
and viola! everything works for me, at least with my RSS feed of https://www.takimag.com/feed/.
I have found that this hack works for other cloudflare protected sites and on the entire site at that (i.e. not just some XML data feed).

tbh i have never made a pull request or even really used github, but can try to do that later on. Keen to hear if this helps anyone else.

**Comment by keresztg** [comment]
@lukerg I can confirm that for my case `curl --no-alpn -A 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:92.0) Gecko/20100101 Firefox/92.0' https://media1.hu/feed/` downloads the feed and `--no-alpn` alone is not enough.

But unfortunately this doesn't work for the case of @squromiv... I still get the usual 'Please turn JavaScript on and reload the page.'

**Comment by squromiv** [comment]
>  I still get the usual 'Please turn JavaScript on and reload the page.'

Yep. It is my pain.

**Comment by Ducky6944** [comment]
`curl --no-alpn -A 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:92.0) Gecko/20100101 Firefox/92.0' https://linuxize.com/index.xml` as mentioned by @keresztg is enough to resolve the issue on my end. I'm not familiar with `--no-alpn` but could be maybe added as an option or implemented in some way. `linuxize.com` is cloudflarred. I did email them about it we will see if they respond.

**Comment by rmdes** [comment]
Damn...I need this so much right now, any news on getting this working ?

**Comment by rmdes** [comment]
Tried --no-alpn on this https://kyivindependent.com/feed/ but it does not work..

**Comment by yue-dongchen** [comment]
@rmdes unrelated to the original issue, but this might be a sensible workaround: use https://github.com/leafac/kill-the-newsletter to convert their newsletter into a feed. I've been following the Kyiv Independent in the past few days as well, but haven't yet tried this tool.

**Comment by rmdes** [comment]
Thanks for the suggestion, I tried KTN for Kyiv Independent, but it's the same, feed cannot be generated because of cloudflare. 
So sad to see Internet getting increasingly jailed :/
